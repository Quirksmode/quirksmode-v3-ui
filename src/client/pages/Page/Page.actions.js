import { push } from 'connected-react-router';

export const FETCH_PAGE_REQUEST = 'fetch_page_request';
export const FETCH_PAGE_SUCCESS = 'fetch_page_success';
export const FETCH_PAGE_ERROR = 'fetch_page_error';
export const fetchPageData = (slug, href = null) => async (
  dispatch,
  getState,
  api
) => {
  // Set the loading state
  dispatch({
    type: FETCH_PAGE_REQUEST,
    payload: slug
  });

  const res = await api.get(`quirksmode/v1/pages/${slug}`);

  dispatch({
    type: FETCH_PAGE_SUCCESS,
    payload: res
  });

  // If a href is passed, navigate to it
  if (href) {
    dispatch(push(href));
  }
};
