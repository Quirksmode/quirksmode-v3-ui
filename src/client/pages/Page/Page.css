.Page {
  border-bottom: 3px solid var(--color-grey-bg);
  overflow: hidden;

  &__section {
    position: relative;
    width: 100%;

    &--greyFade {
      background: linear-gradient(to bottom, var(--color-white) 0%, var(--color-light-grey-fade) 100%);
    }

    &--darkFull {
      background: var(--color-dark-full);
      border: 1px solid var(--color-black);
      border-width: 1px 0 2px;
      box-shadow: inset 0 2px 3px 0 var(--color-black);
    }

    &--whiteOverlay {
      background-color: var(--color-white);
      z-index: var(--z-index-Page__section--whiteOverlay);
    }

    &--withFilter {
      overflow: hidden;
    }
  }

  &__hero {
    overflow: hidden;

    h1 {
      position: relative;
      z-index: var(--z-index-Page__hero__h1);

      @media (--custom-650) {
        width: 51%;
      }
    }

    .Breadcrumbs {
      margin-bottom: 20px;

      @media (--custom-650) {
        margin-bottom: 0;
      }
    }

    &__imgWrap {
      border-radius: 3px;
      bottom: -140px;
      box-shadow: 3px 3px 3px var(--color-sb-grey);
      position: absolute;

      @media (--custom-650) {
        right: 44px;
        transform: rotate(25deg);
      }

      @media (--custom-1000) {
        bottom: -105px;
        width: 392px;
      }

      img {
        width: 100%;
      }
    }
  }

  &__sectionInner {
    padding: 40px 0;
    position: relative;

    @media (--mobile-landscape) {
      padding: 60px 0;
    }

    @media (--tablet-portrait) {
      padding: 80px 0;
    }

    &--dark-arrow {
      &::before {
        display: none;

        @media (--custom-650) {
          background: transparent url(../img/arrow-projectIndicator_2x.png) 0 0;
          background-size: 54px 33px;
          content: '';
          display: block;
          height: 33px;
          left: 5%;
          position: absolute;
          top: -28px;
          width: 54px;
        }
      }
    }
  }

  &__headingWrap {
    margin-bottom: 40px;

    h2 {
      color: var(--color-qm-heading);
      margin-bottom: 0;
    }
  }

  &__content {
    p,
    blockquote,
    ul,
    ol,
    .aspectWrap {
      margin-bottom: 40px;

      &:last-child {
        margin: 0;
      }
    }

    ul {
      list-style-type: disc;
      padding-left: 25px;
    }

    li {
      margin-bottom: 5px;

      &:last-child {
        margin: 0;
      }
    }

    p,
    li {
      font-size: 1.5rem;
      line-height: 2.2rem;

      @media (--mobile-landscape) {
        font-size: 1.6rem;
        line-height: 2.4rem;
      }

      @media (--tablet-portrait) {
        font-size: 1.7rem;
        line-height: 2.5rem;
      }

      &.Page__content__introText {
        color: var(--color-black);
        font-size: 1.7rem;
        line-height: 2.5rem;

        @media (--mobile-landscape) {
          font-size: 1.9rem;
          line-height: 2.8rem;
        }

        @media (--tablet-portrait) {
          font-size: 2rem;
          line-height: 3rem;
        }
      }
    }

    h2 {
      font-size: 4rem;
      line-height: 5rem;
    }

    h3 {
      color: var(--color-emperor);
      font-size: 1.6rem;
      line-height: 2.2rem;
      margin-bottom: 10px;

      @media (--mobile-landscape) {
        font-size: 1.7rem;
        line-height: 2.3rem;
        margin-bottom: 15px;
      }

      @media (--tablet-portrait) {
        font-size: 1.8rem;
        line-height: 2.4rem;
        margin-bottom: 20px;
      }
    }

    blockquote {
      border: 1px solid var(--color-alto);
      border-width: 1px 0;
      padding: 20px;

      p {
        display: inline;
        font-size: 4rem;
        font-weight: 300;
        line-height: 5.4rem;
      }

      &::before,
      &::after {
        font-size: 40px;
        line-height: 1;
        position: relative;
        top: 7px;
      }

      &::before {
        content: '\201C';
        margin: 0 6px 0 -23px;
      }

      &::after {
        content: '\201D';
      }
    }

    iframe {
      width: 100%;
    }
  }
}
