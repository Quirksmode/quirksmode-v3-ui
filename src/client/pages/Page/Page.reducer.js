import {
  FETCH_PAGE_REQUEST,
  FETCH_PAGE_SUCCESS,
  FETCH_PAGE_ERROR
} from './Page.actions';

const initialState = {
  content: {
    title: '',
    pageContent: '',
    contentBlocks: []
  },
  metadata: {},
  loading: false,
  error: false
};

export default (state = initialState, action) => {
  switch (action.type) {
    case FETCH_PAGE_REQUEST:
      return {
        ...state,
        loading: true,
        error: false
      };
    case FETCH_PAGE_SUCCESS:
      return {
        ...state,
        content: action.payload.data.content,
        metadata: action.payload.data.metadata,
        loading: false,
        error: false
      };
    case FETCH_PAGE_ERROR:
      return {
        ...initialState,
        loading: false,
        error: true
      };
    default:
      return state;
  }
};
