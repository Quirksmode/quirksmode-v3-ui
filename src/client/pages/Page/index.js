import React from 'react';
import loadable from '@loadable/component';
import Loading from 'components/Loading/Loading';
import ErrorBoundary from 'components/ErrorBoundary/ErrorBoundary';

const Page = loadable(() => import(/* webpackPrefetch: true */ './Page'), {
  fallback: <Loading />
});

const WrappedPage = props => (
  <ErrorBoundary>
    <Page { ...props } />
  </ErrorBoundary>
);

export default WrappedPage;
