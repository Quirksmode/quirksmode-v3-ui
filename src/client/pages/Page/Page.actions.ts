import { push } from 'connected-react-router';
import axios from 'axios';
import { PageData } from './Page.types';
import {
  fetchPageRequest,
  fetchPageSuccess,
  fetchPageError,
} from './Page.reducer';
import { setLinkLoading, setLinkLoaded } from '../../App.actions';
import { AppThunk } from 'client/redux/types';

export const fetchPageData = (
  slug: string,
  href: string = null
): AppThunk => async (dispatch) => {
  // Set the loading state
  dispatch(fetchPageRequest());

  // If href, dispatch custom Link Loader Functionality
  if (href) {
    dispatch(setLinkLoading(href));
  }

  try {
    const res = await axios.get<PageData>(
      `https://cms.quirksmode.co.uk/wp-json/quirksmode/v1/pages/${slug}`
    );

    dispatch(fetchPageSuccess(res.data));

    // If href, navigate to it
    if (href) {
      dispatch(setLinkLoaded());
      dispatch(push(href));
    }
  } catch {
    dispatch(fetchPageError());
  }
};
