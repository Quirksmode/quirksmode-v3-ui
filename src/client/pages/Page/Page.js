import React, {
  useEffect
} from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Breadcrumbs from 'components/Breadcrumbs/Breadcrumbs';
import Meta from 'components/Meta/Meta';
import ContentBlocks from 'components/ContentBlocks/ContentBlocks';
import { fetchPageData } from './Page.actions';

/**
 * Description
 *
 * @name Page
 * @param  {object} props.cv []
 */
const Page = ({
  fetchPageDataAction,
  match,
  title,
  slug,
  contentBlocks,
  metadata
}) => {
  useEffect(() => {
    if (slug !== match.params.slug) {
      fetchPageDataAction(match.params.slug);
    }
  }, [fetchPageDataAction, match.params.slug, slug]);

  return (
    <div className="page">
      <Meta { ...metadata } />
      <section className="Page__section Page__section--greyFade Page__hero">
        <div className="Page__sectionInner grid">
          <h1>{ title }</h1>
          <Breadcrumbs>
            <span className="Breadcrumbs__divider">&gt;</span>
            <span className="Breadcrumbs__active">{ title }</span>
          </Breadcrumbs>
        </div>
      </section>
      <ContentBlocks contentBlocks={ contentBlocks } />
    </div>
  );
};

Page.propTypes = {
  fetchPageDataAction: PropTypes.func,
  match: PropTypes.object,
  title: PropTypes.string,
  slug: PropTypes.string,
  contentBlocks: PropTypes.array,
  metadata: PropTypes.object
};

const mapStateToProps = ({ page }) => ({
  title: page.title,
  slug: page.slug,
  content: page.content,
  contentBlocks: page.contentBlocks,
  metadata: page.metadata
});

const mapDispatchToProps = dispatch => ({
  fetchPageDataAction: (...args) => dispatch(fetchPageData(...args))
});

export default connect(mapStateToProps, mapDispatchToProps)(Page);
