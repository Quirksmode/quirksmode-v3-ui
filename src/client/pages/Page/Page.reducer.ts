import { PayloadAction, createSlice } from '@reduxjs/toolkit';
import { PageState, PageData } from './Page.types';

export const initialState: PageState = {
  content: {
    title: '',
    slug: '',
    pageContent: '',
    contentBlocks: [],
  },
  metadata: null,
  loading: false,
  error: false,
};

const page = createSlice({
  name: 'page',
  initialState,
  reducers: {
    fetchPageRequest: (state) => {
      state.loading = true;
      state.error = false;
    },
    fetchPageSuccess: (state, { payload }: PayloadAction<PageData>) => {
      state.content = payload.content;
      state.metadata = payload.metadata;
      state.loading = false;
      state.error = false;
    },
    fetchPageError: (state) => {
      state.loading = false;
      state.error = true;
    },
  },
});

export default page.reducer;
export const {
  fetchPageRequest,
  fetchPageSuccess,
  fetchPageError,
} = page.actions;
