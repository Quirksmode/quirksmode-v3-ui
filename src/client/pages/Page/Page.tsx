import React, { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import Breadcrumbs from 'components/Breadcrumbs/Breadcrumbs';
import Meta from 'components/Meta/Meta';
import PageWrapper from 'components/PageWrapper/PageWrapper';
import ContentBlocks from 'components/ContentBlocks/ContentBlocks';
import { fetchPageData } from './Page.actions';
import { useTypedSelector } from 'client/redux/types';
import { useParams } from 'react-router-dom';
import { RouteParams } from './Page.types';

/**
 * Default Page
 */
const Page: React.FC = () => {
  const dispatch = useDispatch();
  const params: RouteParams = useParams();
  const page = useTypedSelector((state) => state.page);

  const { content, metadata, loading, error } = page;
  const { title, contentBlocks } = content;

  // Fetch the Page Data via Redux using the routes slug param
  useEffect(() => {
    dispatch(fetchPageData(params.slug));
  }, [params.slug]);

  return (
    <PageWrapper error={error} loading={loading}>
      <div className="page">
        {metadata && <Meta {...metadata} />}
        <section className="Page__section Page__section--greyFade Page__hero">
          <div className="Page__sectionInner grid">
            <h1>{title}</h1>
            <Breadcrumbs>
              <span className="Breadcrumbs__divider">&gt;</span>
              <span className="Breadcrumbs__active">{title}</span>
            </Breadcrumbs>
          </div>
        </section>
        <ContentBlocks contentBlocks={contentBlocks} />
      </div>
    </PageWrapper>
  );
};

export default Page;
