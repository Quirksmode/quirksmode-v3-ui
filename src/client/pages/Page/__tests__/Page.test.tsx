import {
  findByTestAttr,
  mountRender,
  reduxHelper,
} from 'client/tests/testUtils';
import moxios from 'moxios';
import Page from '../Page';
import AppStub from 'client/__tests__/app.stub.json';
import PageStub from './Page.stub.json';
import PageWrapper from 'client/components/PageWrapper/PageWrapper';

describe('Page', () => {
  let wrapper;
  const props = {};
  const state = {
    app: AppStub,
    page: PageStub,
  };
  const routeData = {
    initialEntries: ['/accessibility'],
    path: '/:slug',
  };

  beforeEach(() => {
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  it('matches snapshot and renders correctly with data', () => {
    const { ReduxWrapper } = reduxHelper(Page, props, state, routeData);
    wrapper = mountRender(ReduxWrapper);
    expect(wrapper.find(PageWrapper).length).toEqual(1);
    expect(findByTestAttr(wrapper, 'Breadcrumbs').length).toEqual(1);
    expect(wrapper.html).toMatchSnapshot();
  });

  it('renders correctly when there is no data', () => {
    const { ReduxWrapper } = reduxHelper(Page, props, {}, routeData);
    wrapper = mountRender(ReduxWrapper);
    expect(wrapper.find(PageWrapper).length).toEqual(1);
    expect(wrapper.html).toMatchSnapshot();
  });

  it('runs the fetchPageData action on page load', () => {
    const newState = {
      ...state,
      page: {
        ...state.page,
        content: {
          ...state.page.content,
          title: '',
        },
      },
    };
    const { dispatch, ReduxWrapper } = reduxHelper(
      Page,
      props,
      newState,
      routeData
    );
    wrapper = mountRender(ReduxWrapper);
    expect(dispatch).toHaveBeenCalledTimes(1);
  });

  it('runs the correct actions when the Page data is successfully fetched', async (done) => {
    const { getActions, ReduxWrapper } = reduxHelper(
      Page,
      props,
      {},
      routeData
    );
    wrapper = mountRender(ReduxWrapper);

    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request
        .respondWith({
          status: 200,
          response: PageStub,
        })
        .then(() => {
          const expected = [
            {
              type: 'page/fetchPageRequest',
            },
            {
              type: 'page/fetchPageSuccess',
              payload: PageStub,
            },
          ];
          const actual = getActions();
          expect(actual).toEqual(expected);
          done();
        });
    });
  });

  it('runs the correct actions when the Page data is not successfully fetched', async (done) => {
    const { getActions, ReduxWrapper } = reduxHelper(
      Page,
      props,
      {},
      routeData
    );
    wrapper = mountRender(ReduxWrapper);

    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request
        .respondWith({
          status: 500,
        })
        .then(() => {
          const expected = [
            {
              type: 'page/fetchPageRequest',
            },
            {
              type: 'page/fetchPageError',
            },
          ];
          const actual = getActions();
          expect(actual).toEqual(expected);
          done();
        });
    });
  });
});
