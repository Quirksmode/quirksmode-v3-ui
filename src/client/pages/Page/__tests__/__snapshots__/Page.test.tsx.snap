// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page matches snapshot and renders correctly with data 1`] = `[Function]`;

exports[`Page renders correctly when there is no data 1`] = `[Function]`;
